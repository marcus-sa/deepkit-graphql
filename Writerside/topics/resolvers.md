<title>Resolvers</title>

Start typing here...

<chapter title="Object Types">
<p>
<a href="https://graphql.org/graphql-js/object-types/">Object Types</a> are written using TypeScript.
</p>
<tabs>
<tab title="Author">
</tab>
<tab title="Post">
</tab>
</tabs>
</chapter>
