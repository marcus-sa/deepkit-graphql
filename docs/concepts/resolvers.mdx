## Queries

## Mutations

## Subscriptions

## Fields

## Context

## Middleware

## Parent
