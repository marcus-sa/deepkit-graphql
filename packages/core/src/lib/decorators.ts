/* eslint-disable functional/immutable-data,functional/prefer-readonly-type,@typescript-eslint/typedef */
import { ClassType } from '@deepkit/core';
import {
  ClassDecoratorFn,
  createClassDecoratorContext,
  createPropertyDecoratorContext,
  DecoratorAndFetchSignature,
  DualDecorator,
  ExtractApiDataType,
  ExtractClass,
  mergeDecorator,
  PropertyDecoratorResult,
  ReceiveType,
  ReflectionClass,
  ReflectionKind,
  resolveReceiveType,
  resolveRuntimeType,
  Type,
  TypeClass,
  TypeObjectLiteral,
  UnionToIntersection,
} from '@deepkit/type';

import { requireTypeName, unwrapPromiseLikeType } from './types-builder';

export const typeResolvers = new Map<string, ClassType>();

/*export function isValidMethodReturnType(
  classType: ClassType,
  methodName: string,
): boolean {
  const resolverType = resolveRuntimeType(classType);
  const reflectionClass = ReflectionClass.from(resolverType);
  const method = reflectionClass.getMethod(methodName);
  let returnType = method.getReturnType();
  returnType = unwrapPromiseLikeType(returnType);

  return returnType.kind !== ReflectionKind.union
    ? returnType.kind === ReflectionKind.objectLiteral ||
        returnType.kind === ReflectionKind.class
    : returnType.types.some(
        type =>
          type.kind === ReflectionKind.objectLiteral ||
          type.kind === ReflectionKind.class ||
          type.kind === ReflectionKind.null ||
          type.kind === ReflectionKind.undefined,
      );
}*/

class GraphQLResolver {
  type?: TypeClass | TypeObjectLiteral;

  classType: ClassType;

  readonly mutations = new Map<string, GraphQLMutationMetadata>();

  readonly queries = new Map<string, GraphQLQueryMetadata>();

  readonly resolveFields = new Map<string, GraphQLFieldMetadata>();

  readonly checks = new Set<(decorator: GraphQLResolverDecorator) => void>();
}

class GraphQLResolverDecorator {
  t = new GraphQLResolver();

  onDecorator(classType: ClassType) {
    this.t.classType = classType;
  }

  resolver<T>(type?: ReceiveType<T>) {
    try {
      type = resolveReceiveType(type);
      if (
        type.kind !== ReflectionKind.class &&
        type.kind !== ReflectionKind.objectLiteral
      ) {
        throw new Error(
          'Only classes and interfaces are supported by @graphql.resolver<T>()',
        );
      }
      this.t.type = type;
      const typeName = requireTypeName(type);
      typeResolvers.set(typeName, this.t.classType);
    } catch (err) {
      if (
        err instanceof Error &&
        err.message !==
          'No type information received. Is deepkit/type correctly installed?'
      ) {
        throw err;
      }
    }

    this.t.checks.forEach(check => check(this));
  }

  addMutation(property: string, mutation: GraphQLMutationMetadata) {
    this.t.mutations.set(property, mutation);
  }

  addQuery(property: string, query: GraphQLQueryMetadata) {
    this.t.queries.set(property, query);
  }

  addResolveField(property: string, field: GraphQLFieldMetadata) {
    this.t.resolveFields.set(property, field);
  }

  addCheck(check: (decorator: GraphQLResolverDecorator) => void) {
    this.t.checks.add(check);
  }
}

interface GraphQLQueryOptions {
  description?: string;
  deprecationReason?: string;
}

export class GraphQLQueryMetadata implements GraphQLQueryOptions {
  name: string;
  classType: ClassType;
  description?: string;
  deprecationReason?: string;
  readonly checks = new Set<(decorator: GraphQLResolverDecorator) => void>();
}

class GraphQLQueryDecorator {
  t = new GraphQLQueryMetadata();

  onDecorator(classType: ClassType, property: string | undefined) {
    if (!property) return;
    this.t.name = property;
    this.t.classType = classType;
    gqlResolverDecorator.addQuery(property, this.t)(classType);

    this.t.checks.forEach(check =>
      gqlResolverDecorator.addCheck(check)(classType),
    );
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  query(options?: GraphQLQueryOptions) {
    this.t.description = options?.description;
    this.t.deprecationReason = options?.deprecationReason;

    /*this.t.checks.add(() => {
      if (!isValidMethodReturnType(this.t.classType, this.t.name)) {
        throw new Error(
          'Only classes and interfaces are supported as return types for methods decorated by @graphql.query()',
        );
      }
    });*/
  }
}

export const gqlQueryDecorator: PropertyDecoratorResult<
  typeof GraphQLQueryDecorator
> = createPropertyDecoratorContext(GraphQLQueryDecorator);

interface GraphQLMutationOptions {
  description?: string;
  deprecationReason?: string;
}

export class GraphQLMutationMetadata implements GraphQLMutationOptions {
  name: string;
  classType: ClassType;
  description?: string;
  deprecationReason?: string;
  readonly checks = new Set<(decorator: GraphQLResolverDecorator) => void>();
}

class GraphQLMutationDecorator {
  t = new GraphQLMutationMetadata();

  onDecorator(classType: ClassType, property: string | undefined) {
    if (!property) return;
    this.t.name = property;
    this.t.classType = classType;
    gqlResolverDecorator.addMutation(property, this.t)(classType);

    this.t.checks.forEach(check =>
      gqlResolverDecorator.addCheck(check)(classType),
    );
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  mutation(options?: GraphQLMutationOptions) {
    this.t.description = options?.description;
    this.t.deprecationReason = options?.deprecationReason;

    /*this.t.checks.add(() => {
      if (!isValidMethodReturnType(this.t.classType, this.t.name)) {
        throw new Error(
          'Only classes and interfaces are supported as return types for methods decorated by @graphql.mutation()',
        );
      }
    });*/
  }
}

export const gqlMutationDecorator: PropertyDecoratorResult<
  typeof GraphQLMutationDecorator
> = createPropertyDecoratorContext(GraphQLMutationDecorator);

export class GraphQLFieldMetadata {
  property: string;
  classType: ClassType;
  name: string;

  readonly checks = new Set<(decorator: GraphQLResolverDecorator) => void>();
}

class GraphQLFieldDecorator {
  t = new GraphQLFieldMetadata();

  onDecorator(classType: ClassType, property: string | undefined) {
    if (!property) return;
    this.t.property = property;
    this.t.name ||= property;
    this.t.classType = classType;
    gqlResolverDecorator.addResolveField(property, this.t)(classType);

    this.t.checks.forEach(check =>
      gqlResolverDecorator.addCheck(check)(classType),
    );
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  resolveField(name?: string) {
    if (name) {
      this.t.name = name;
    }

    this.t.checks.add(resolverDecorator => {
      if (!resolverDecorator.t.type) {
        throw new Error(
          'Can only resolve fields for resolvers with a type @graphql.resolver<T>()',
        );
      }

      const resolverType = resolveRuntimeType(this.t.classType);
      const reflectionClass = ReflectionClass.from(resolverType);

      if (!reflectionClass.hasMethod(this.t.name)) {
        const typeName = requireTypeName(resolverDecorator.t.type);
        throw new Error(
          `No field ${this.t.name} found on type ${typeName} for field resolver method ${this.t.property} on resolver ${this.t.classType.name}`,
        );
      }
    });
  }
}

export const gqlFieldDecorator: PropertyDecoratorResult<
  typeof GraphQLFieldDecorator
> = createPropertyDecoratorContext(GraphQLFieldDecorator);

//this workaround is necessary since generic functions (necessary for response<T>) are lost during a mapped type and changed ReturnType
// eslint-disable-next-line @typescript-eslint/ban-types
type GraphQLResolverFluidDecorator<T, D extends Function> = {
  [name in keyof T]: name extends 'resolver'
    ? <For>(type?: ReceiveType<For>) => D & GraphQLResolverFluidDecorator<T, D>
    : T[name] extends (...args: infer K) => any
    ? (...args: K) => D & GraphQLResolverFluidDecorator<T, D>
    : D &
        GraphQLResolverFluidDecorator<T, D> & { _data: ExtractApiDataType<T> };
};

type GraphQLResolverClassDecoratorResult = GraphQLResolverFluidDecorator<
  ExtractClass<typeof GraphQLResolverDecorator>,
  ClassDecoratorFn
> &
  DecoratorAndFetchSignature<typeof GraphQLResolverDecorator, ClassDecoratorFn>;

export const gqlResolverDecorator: GraphQLResolverClassDecoratorResult =
  createClassDecoratorContext(GraphQLResolverDecorator);

//this workaround is necessary since generic functions are lost during a mapped type and changed ReturnType
type GraphQLMerge<U> = {
  [K in keyof U]: K extends 'resolver'
    ? <For>(type?: ReceiveType<For>) => ClassDecoratorFn & U
    : U[K] extends (...a: infer A) => infer R
    ? R extends DualDecorator
      ? (...a: A) => ClassDecoratorFn & R & U
      : (...a: A) => R
    : never;
};

type MergedGraphQL<T extends any[]> = GraphQLMerge<
  Omit<UnionToIntersection<T[number]>, '_fetch' | 't'>
>;

export type GraphQLDecorator = ClassDecoratorFn &
  GraphQLResolverFluidDecorator<GraphQLResolverDecorator, ClassDecoratorFn>;

export type MergedGraphQLDecorator = Omit<
  MergedGraphQL<
    [
      typeof gqlResolverDecorator,
      typeof gqlFieldDecorator,
      typeof gqlMutationDecorator,
      typeof gqlQueryDecorator,
    ]
  >,
  'addMutation' | 'addQuery' | 'addResolveField' | 'onDecorator' | 'addCheck'
>;

// TODO: subscriptions
export const graphql: MergedGraphQLDecorator = mergeDecorator(
  gqlResolverDecorator,
  gqlFieldDecorator,
  gqlMutationDecorator,
  gqlQueryDecorator,
) as any as MergedGraphQLDecorator;
