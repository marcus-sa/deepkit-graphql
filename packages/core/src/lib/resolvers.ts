import { InjectorModule } from '@deepkit/injector';
import { ClassType } from '@deepkit/core';

export interface Resolver<T> {
  readonly controller: ClassType<T>;
  readonly module?: InjectorModule;
}

export class Resolvers extends Set<Resolver<unknown>> {}
