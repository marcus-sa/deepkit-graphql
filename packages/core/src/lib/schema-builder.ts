import { ReceiveType, resolveReceiveType, Type } from '@deepkit/type';
import { ClassType } from '@deepkit/core';
import {
  GraphQLFieldConfigMap,
  GraphQLNamedType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
} from 'graphql';

import { Resolvers } from './resolvers';
import { TypesBuilder } from './types-builder';
import { gqlResolverDecorator } from './decorators';

export interface SchemaBuilderOptions {
  readonly resolvers: Resolvers;
  readonly inputTypes?: readonly Type[];
  readonly outputTypes?: readonly Type[];
}

export function buildSchema(options: SchemaBuilderOptions): GraphQLSchema {
  return new SchemaBuilder(options).buildSchema();
}

export class SchemaBuilder {
  private readonly inputTypes = new Set<Type>(this.options.inputTypes);

  private readonly outputTypes = new Set<Type>(this.options.outputTypes);

  private readonly typesBuilder = new TypesBuilder(this.options.resolvers);

  constructor(private readonly options: SchemaBuilderOptions) {}

  private buildInputTypes(): readonly GraphQLNamedType[] {
    return [...this.inputTypes].map(type =>
      this.typesBuilder.createNamedInputType(type),
    );
  }

  private buildOutputTypes(): readonly GraphQLNamedType[] {
    return [...this.outputTypes].map(type =>
      this.typesBuilder.createNamedOutputType(type),
    );
  }

  generateMutationResolverFields(): GraphQLFieldConfigMap<unknown, unknown> {
    return [...this.options.resolvers.instances].reduce<
      GraphQLFieldConfigMap<unknown, unknown>
    >(
      (fields, instance) => ({
        // TODO: validate that fields don't override each other
        ...fields,
        ...this.typesBuilder.generateMutationResolverFields(instance),
      }),
      {},
    );
  }

  generateQueryResolverFields(): GraphQLFieldConfigMap<unknown, unknown> {
    return [...this.options.resolvers.instances].reduce<
      GraphQLFieldConfigMap<unknown, unknown>
    >(
      (fields, instance) => ({
        // TODO: validate that fields don't override each other
        ...fields,
        ...this.typesBuilder.generateQueryResolverFields(instance),
      }),
      {},
    );
  }

  private buildRootMutationType(): GraphQLObjectType | undefined {
    const classTypes = [...this.options.resolvers.classTypes];

    const someMutations = classTypes.some(classType =>
      this.hasMutationResolvers(classType),
    );
    if (!someMutations) return;

    return new GraphQLObjectType({
      name: 'Mutation',
      fields: () => this.generateMutationResolverFields(),
    });
  }

  /*private buildRootSubscriptionType(
    resolvers: Function[],
  ): GraphQLObjectType | undefined {
    const subscriptionsHandlers = this.filterHandlersByResolvers(
      getMetadataStorage().subscriptions,
      resolvers,
    );
    if (subscriptionsHandlers.length === 0) {
      return undefined;
    }

    return new GraphQLObjectType({
      name: 'Subscription',
      fields: this.generateSubscriptionsFields(subscriptionsHandlers),
    });
  }*/

  hasQueryResolvers(classType: ClassType): boolean {
    const resolver = gqlResolverDecorator._fetch(classType);
    return !!resolver?.queries.size;
  }

  hasMutationResolvers(classType: ClassType): boolean {
    const resolver = gqlResolverDecorator._fetch(classType);
    return !!resolver?.mutations.size;
  }

  hasFieldResolvers(classType: ClassType): boolean {
    const resolver = gqlResolverDecorator._fetch(classType);
    return !!resolver?.resolveFields.size;
  }

  private buildRootQueryType(): GraphQLObjectType | undefined {
    const classTypes = [...this.options.resolvers.classTypes];

    const someQueries = classTypes.some(classType =>
      this.hasQueryResolvers(classType),
    );
    if (!someQueries) {
      return new GraphQLObjectType({
        name: 'Query',
        fields: {
          _dummy: { type: GraphQLString },
        },
      });
    }

    return new GraphQLObjectType({
      name: 'Query',
      fields: () => this.generateQueryResolverFields(),
    });
  }

  addInputType<T>(type?: ReceiveType<T>): void {
    type = resolveReceiveType(type);

    if (!this.inputTypes.has(type)) {
      this.inputTypes.add(type);
    }
  }

  addOutputType<T>(type?: ReceiveType<T>): void {
    type = resolveReceiveType(type);

    if (!this.outputTypes.has(type)) {
      this.outputTypes.add(type);
    }
  }

  buildSchema(): GraphQLSchema {
    const mutation = this.buildRootMutationType();
    const query = this.buildRootQueryType();
    const types = [...this.buildInputTypes(), ...this.buildOutputTypes()];

    return new GraphQLSchema({
      query,
      mutation,
      types,
    });
  }
}
