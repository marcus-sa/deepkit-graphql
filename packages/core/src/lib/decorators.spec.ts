import { graphql as executeGraphQL } from 'graphql';
import {
  assert,
  BackReference,
  cast,
  entity,
  integer,
  MinLength,
  PositiveNoZero,
  PrimaryKey,
  Reference,
  uuid,
  UUID,
} from '@deepkit/type';

import { graphql } from './decorators';
import { Context, Parent } from './types-builder';
import { buildSchema } from './schema-builder';
import { Resolvers } from './resolvers';

/*test('invalid return type for mutation', () => {
  expect(() => {
    @graphql.resolver()
    class TestResolver {
      @graphql.mutation()
      mutation(): string {
        return '';
      }
    }
  }).toThrowErrorMatchingSnapshot();
});*/

/*test('invalid return type for query', () => {
  expect(() => {
    @graphql.resolver()
    class TestResolver {
      @graphql.query()
      query(): string {
        return '';
      }
    }
  }).toThrowErrorMatchingSnapshot();
});*/

test('mutation', async () => {
  interface User {
    readonly id: integer & PositiveNoZero;
    readonly username: string;
  }

  interface CreateUserData {
    readonly username: string;
  }

  @graphql.resolver()
  class UserResolver {
    // eslint-disable-next-line @typescript-eslint/require-await
    @graphql.mutation({ description: 'Create user' })
    async createUser(data: CreateUserData): Promise<User | null> {
      return { id: 1, ...data };
    }
  }

  const resolvers = new Resolvers([new UserResolver()]);

  const schema = buildSchema({ resolvers });

  await expect(
    executeGraphQL({
      schema,
      source: `mutation { createUser(data: { username: "Test" }) { id } }`,
    }),
  ).resolves.toMatchSnapshot();
});

test('query', async () => {
  interface User {
    readonly id: integer & PositiveNoZero;
  }

  @graphql.resolver()
  class UserResolver {
    // eslint-disable-next-line @typescript-eslint/require-await
    @graphql.query()
    async getUser(id: User['id']): Promise<User | null> {
      return { id };
    }
  }

  const resolvers = new Resolvers([new UserResolver()]);

  const schema = buildSchema({ resolvers });

  await expect(
    executeGraphQL({
      schema,
      source: `{ getUser(id: 1) { id } }`,
      rootValue: {},
    }),
  ).resolves.toMatchSnapshot(); // Can't use .toMatchInlineSnapshot() because of decorators
});

test('mutation args validation', async () => {
  interface User {
    readonly username: string;
  }

  interface CreateUserData {
    readonly username: string & MinLength<6>;
  }

  @graphql.resolver()
  class UserResolver {
    // eslint-disable-next-line @typescript-eslint/require-await
    @graphql.mutation()
    async createUser(data: CreateUserData): Promise<User | null> {
      return data;
    }
  }

  const resolvers = new Resolvers([new UserResolver()]);

  const schema = buildSchema({ resolvers });

  await expect(
    executeGraphQL({
      source: `mutation { createUser(data: { username: "Test" }) { username } }`,
      schema,
    }),
  ).resolves.toMatchSnapshot(); // Can't use .toMatchInlineSnapshot() because of decorators
});

test('Context', async () => {
  interface TestCtx {
    readonly version: string;
  }

  const testCtx: TestCtx = {
    version: '1.0.0',
  };

  interface Info {
    readonly version: string;
  }

  @graphql.resolver()
  class TestResolver {
    @graphql.query()
    info(ctx: Context<TestCtx>): Info {
      expect(ctx).toBe(testCtx);
      return { version: ctx.version };
    }
  }

  const resolvers = new Resolvers([new TestResolver()]);

  const schema = buildSchema({ resolvers });

  await expect(
    executeGraphQL({
      contextValue: testCtx,
      source: `
        {
          info {
            version
          }
        }
      `,
      schema,
    }),
  ).resolves.toMatchSnapshot(); // Can't use .toMatchInlineSnapshot() because of decorators
});

describe('resolveField', () => {
  test('parent', async () => {
    @entity.name('post')
    class Post {
      readonly id: UUID & PrimaryKey = uuid();

      readonly author?: User & Reference;
    }

    @entity.name('user')
    class User {
      readonly id: UUID & PrimaryKey = uuid();
      readonly posts?: readonly Post[] & BackReference;
    }

    @graphql.resolver<Post>()
    class PostResolver {
      @graphql.resolveField()
      async author(post: Parent<Post>): Promise<User> {
        console.log(post);
        return cast<User>({});
      }
    }

    @graphql.resolver<User>()
    class UserResolver {
      @graphql.resolveField()
      posts(user: Parent<User>): readonly Post[] | undefined {
        assert<User>(user);
        return [];
      }

      @graphql.query()
      getUser(id: UUID): User {
        return cast<User>({ id });
      }
    }

    const resolvers = new Resolvers([new UserResolver(), new PostResolver()]);

    const schema = buildSchema({ resolvers });

    await expect(
      executeGraphQL({
        schema,
        source: `
          {
            getUser(id: "9f617521-b9c2-4ab9-a339-3c551c799027") {
              id
              posts {
                id
              }
            }
          }
        `,
      }),
    ).resolves.toMatchSnapshot();
  });

  test('resolves fields', async () => {
    interface Post {
      readonly id: UUID;
      readonly author?: User;
    }

    interface User {
      readonly id: UUID;
      readonly posts?: readonly Post[];
    }

    @graphql.resolver<User>()
    class UserResolver {
      @graphql.resolveField()
      posts(): readonly Post[] | undefined {
        return [cast<Post>({ id: '13fe73c3-af3d-4979-bc2d-f66ca1960a76' })];
      }

      @graphql.query()
      getUser(id: UUID): User {
        return cast<User>({ id });
      }
    }

    @graphql.resolver<Post>()
    class PostResolver {
      @graphql.resolveField()
      author(): User | undefined {
        return cast<User>({ id: '398cdf36-e3ac-475c-90aa-c70f99add874' });
      }
    }

    const resolvers = new Resolvers([new UserResolver(), new PostResolver()]);

    const schema = buildSchema({ resolvers });

    await expect(
      executeGraphQL({
        source: `
          {
            getUser(id: "398cdf36-e3ac-475c-90aa-c70f99add874") {
              id
              posts {
                id
                author {
                  id
                }
              }
            }
          }
        `,
        schema,
      }),
    ).resolves.toMatchSnapshot(); // Can't use .toMatchInlineSnapshot() because of decorators
  });

  // FIXME: only works when run alone
  /*test('resolver missing generic', () => {
    interface Post {
      readonly id: string;
      readonly author?: User;
    }

    interface User {
      readonly id: string;
      readonly posts?: readonly Post[];
    }

    expect(() => {
      @graphql.resolver()
      class UserResolver {
        @graphql.resolveField()
        posts(): readonly Post[] | null {
          return [];
        }
      }
    }).toThrowErrorMatchingSnapshot();
  });*/

  test('unknown property', () => {
    interface Post {
      readonly id: string;
      readonly author?: User;
    }

    interface User {
      readonly id: string;
      readonly posts?: readonly Post[];
    }

    expect(() => {
      @graphql.resolver<User>()
      class UserResolver {
        @graphql.resolveField('post')
        posts(): readonly Post[] | null {
          return [];
        }
      }
    }).toThrowErrorMatchingSnapshot();
  });
});
