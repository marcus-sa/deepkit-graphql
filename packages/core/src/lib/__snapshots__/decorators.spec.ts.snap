// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context 1`] = `
{
  "data": {
    "info": {
      "version": "1.0.0",
    },
  },
}
`;

exports[`mutation 1`] = `
{
  "data": {
    "createUser": {
      "id": 1,
    },
  },
}
`;

exports[`mutation args validation 1`] = `
{
  "data": {
    "createUser": null,
  },
  "errors": [
    [GraphQLError: Validation error:
data.username(minLength): Min length is 6],
  ],
}
`;

exports[`query 1`] = `
{
  "data": {
    "getUser": {
      "id": 1,
    },
  },
}
`;

exports[`resolveField parent 1`] = `
{
  "data": {
    "getUser": {
      "id": "9f617521-b9c2-4ab9-a339-3c551c799027",
      "posts": [],
    },
  },
}
`;

exports[`resolveField resolves fields 1`] = `
{
  "data": {
    "getUser": {
      "id": "398cdf36-e3ac-475c-90aa-c70f99add874",
      "posts": [
        {
          "author": {
            "id": "398cdf36-e3ac-475c-90aa-c70f99add874",
          },
          "id": "13fe73c3-af3d-4979-bc2d-f66ca1960a76",
        },
      ],
    },
  },
}
`;

exports[`resolveField unknown property 1`] = `"No field post found on type User for field resolver method posts on resolver UserResolver"`;
